File: coding/leetcode/331-Verify-Preorder-Serialization-of-a-Binary-Tree.go
func isValidSerialization(preorder string) bool {
    num := 1
    for i := 0; i < len(preorder); i++ {
        if preorder[i] == ',' {
            continue
        }
        if num <= 0 {
            return false
        }
        if preorder[i] == '#' {
            num--
        } else {
            num++
            for i < len(preorder) && preorder[i] != '#' {
                i++
            }
        }
    }
    return num == 0
}

File: oi.wiki/leetcode/331-Verify-Preorder-Serialization-of-a-Binary-Tree.cpp
#include <string>
using std::string;
bool isValidSerialization(string preorder) {
    int num = 1;
    for (int i = 0; i < preorder.length(); i++) {
        if (preorder[i] == ',') continue;
        if (num <= 0) return false;
        if (isdigit(preorder[i])) {
            num++;
            while (i + 1 < preorder.length() && isdigit(preorder[i + 1])) i++;
        } else if (preorder[i] == '#') {
            num--;
        }
    }
    return num == 0;
}
